File: server/routes/journalHistory.js
const express = require('express');
const { getDb } = require('../services/db');

const router = express.Router();

// GET /api/journal-history — public (all logged-in users can view)
router.get('/', async (req, res) => {
  const db = getDb();
  if (!db) return res.json({ posts: [] });

  const { data, error } = await db
    .from('journal_posts')
    .select('*')
    .order('posted_at', { ascending: false })
    .limit(200);

  if (error) return res.status(500).json({ error: error.message });
  res.json({ posts: data || [] });
});

// GET /api/journal-history/stats
router.get('/stats', async (req, res) => {
  const db = getDb();
  if (!db) return res.json({ byEntity: [], byModule: [], recentMonths: [] });

  const { data: posts } = await db.from('journal_posts').select('entity, module, total_amount, posted_at, journal_date');

  if (!posts) return res.json({ byEntity: [], byModule: [], recentMonths: [] });

  const byEntity = {};
  const byModule = { csi: { count: 0, total: 0 }, payroll: { count: 0, total: 0 } };
  const byMonth = {};

  for (const p of posts) {
    // by entity
    if (!byEntity[p.entity]) byEntity[p.entity] = { count: 0, total: 0 };
    byEntity[p.entity].count++;
    byEntity[p.entity].total += parseFloat(p.total_amount || 0);

    // by module
    const mod = p.module || 'csi';
    byModule[mod] = byModule[mod] || { count: 0, total: 0 };
    byModule[mod].count++;
    byModule[mod].total += parseFloat(p.total_amount || 0);

    // by month
    const ym = (p.journal_date || p.posted_at || '').slice(0, 7);
    if (ym) {
      if (!byMonth[ym]) byMonth[ym] = { count: 0, total: 0 };
      byMonth[ym].count++;
      byMonth[ym].total += parseFloat(p.total_amount || 0);
    }
  }

  const recentMonths = Object.entries(byMonth)
    .sort(([a], [b]) => b.localeCompare(a))
    .slice(0, 12)
    .map(([month, v]) => ({ month, ...v }));

  res.json({
    byEntity: Object.entries(byEntity).map(([entity, v]) => ({ entity, ...v })).sort((a, b) => b.total - a.total),
    byModule: Object.entries(byModule).map(([mod, v]) => ({ module: mod, ...v })),
    recentMonths,
    totalPosts: posts.length,
    totalAmount: posts.reduce((s, p) => s + parseFloat(p.total_amount || 0), 0),
  });
});

module.exports = router;

File: app/routers/journal_history.py
from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse
from app.services.db import get_db

router = APIRouter(prefix="/api/journal-history")


@router.get("")
async def list_journals(request: Request):
    db = get_db()
    if not db:
        return JSONResponse({"posts": []})
    resp = db.from_("journal_posts").select("*").order("posted_at", desc=True).limit(200).execute()
    return JSONResponse({"posts": resp.data or []})


@router.get("/stats")
async def journal_stats(request: Request):
    db = get_db()
    if not db:
        return JSONResponse({"byEntity": [], "byModule": [], "recentMonths": [], "totalPosts": 0, "totalAmount": 0})

    resp = db.from_("journal_posts").select("entity, module, total_amount, posted_at, journal_date").execute()
    posts = resp.data or []

    by_entity: dict = {}
    by_module: dict = {}
    by_month: dict = {}
    total_amount = 0.0

    for p in posts:
        amount = float(p.get("total_amount") or 0)
        total_amount += amount

        entity = p.get("entity", "")
        if entity not in by_entity:
            by_entity[entity] = {"count": 0, "total": 0.0}
        by_entity[entity]["count"] += 1
        by_entity[entity]["total"] += amount

        mod = p.get("module") or "csi"
        if mod not in by_module:
            by_module[mod] = {"count": 0, "total": 0.0}
        by_module[mod]["count"] += 1
        by_module[mod]["total"] += amount

        ym = (p.get("journal_date") or p.get("posted_at") or "")[:7]
        if ym:
            if ym not in by_month:
                by_month[ym] = {"count": 0, "total": 0.0}
            by_month[ym]["count"] += 1
            by_month[ym]["total"] += amount

    recent_months = sorted(
        [{"month": k, **v} for k, v in by_month.items()],
        key=lambda x: x["month"],
        reverse=True,
    )[:12]

    return JSONResponse({
        "byEntity": sorted([{"entity": k, **v} for k, v in by_entity.items()], key=lambda x: x["total"], reverse=True),
        "byModule": [{"module": k, **v} for k, v in by_module.items()],
        "recentMonths": recent_months,
        "totalPosts": len(posts),
        "totalAmount": total_amount,
    })
